### ISSUE TITLE
A general hook for take full controll of twig

### ISSUE THREAD

**Issue by matheuswf** [simplesamlphp/simplesamlphp#1482]
Hi.

I need to take control of twig for add a custom.css and others thing in loginuserpass.php of module core. But i don't wanna modify original files because if i upgrade the original module, i lost my changes.

I sugest two hooks: One after at class SimpleSAML\XHTML\Template at end of contructor and another in method show() or getContents().

:D

**Comment by tvdijen** [comment]
Hi @matheuswf ! You can use a theme to override existing templates..
The docs may be a little outdated;  https://github.com/simplesamlphp/simplesamlphp/blob/master/docs/simplesamlphp-theming.md

The idea is to create your own module, configure it as 'theme.use' in `config.php` and then in the theme-module you can override the core:loginuserpass.twig template.  You can even add more custom processing to it if you write your own ThemeController.  Would that cover your needs?

**Comment by matheuswf** [comment]
Yes, that's what I needed. Sorry for opening this issue. I really hadn't found that solution and so I opened it.

**Comment by tvdijen** [comment]
No worries! I'll keep it open for a few days .. Ask me anything if you can't figure it out

**Comment by matheuswf** [comment]
Hi.

I use 'usenewui' feature with true value in config. This allow to use PO file for translations and I try to find an way to add custom PO file. In Controller defined by 'theme.controller' I can access &$twig, but not base Template.

**Comment by tvdijen** [comment]
I don't think it's currently possible :(  I guess we could pass some more to the theme controller, like the Localization and Translator

**Comment by matheuswf** [comment]
it would be nice to have in addition to Twig, Localization and Translator or maybe full access to the Template.

**Comment by tvdijen** [comment]
Yes, I'm thinking the latter because it grants indirect access to the other two..
My co-dev mentioned that it should be possible to access Gettext statically, but I think this shouldn't be that hard for our users..

The change itself is easy enough.. I can come up with the change in the next couple of days.. The only downside for you perhaps is that a change in the signature means a BC-break, so it may take a while to become part of a released version.

I know I can use ReflectionMethod to see if the controller implements the old or the new signature, but that used to be a really expensive check, so in this case that's not really an option..
